### PR TITLE
#285 crontab job keys

### PR DESCRIPTION
**Pull request by spiffytech**
## Description

Implements #285: if a cron job would reschedule (due to an error or just not getting picked up), only ever schedule one copy of the job if the crontab includes a `jobKey` parameter.scord discussion/etc as appropriate -->

## Performance impact

Unknown

## Security impact

Unknown

## Checklist

- [ ] My code matches the project's code style and `yarn lint:fix` passes.
- [ ] I've added tests for the new feature, and `yarn test` passes.
- [X] I have detailed the new feature in the relevant documentation.
- [X] I have added this feature to 'Pending' in the `RELEASE_NOTES.md` file (if one exists).
`yarn lint:fix` fails with this:
```
$ yarn lint fix
yarn run v1.22.19
$ yarn prettier:check && eslint --ext .js,.jsx,.ts,.tsx,.graphql . fix
$ prettier --ignore-path .eslintignore --check '**/*.{js,jsx,ts,tsx,graphql,md,json}'
Checking formatting...
All matched files use Prettier code style!

Oops! Something went wrong! :(

ESLint: 8.18.0

No files matching the pattern "fix" were found.
Please check for typing mistakes in the pattern.

error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

`yarn test` from inside the container fails with this:

```
$ docker compose exec app yarn test   
yarn run v1.22.19                                                                                        
$ yarn prepack && depcheck && createdb graphile_worker_test || true && psql -X -v GRAPHILE_WORKER_SCHEMA="${GRAPHILE_WORKER_SCHEMA:-graphile_worker}" -v ON_ERROR_STOP=1 -f __tests__/reset-db.sql graphile_worker_
test && jest -i                                     
$ rm -Rf dist && tsc && chmod +x dist/cli.js
Unused dependencies                                                                                                                                                                                                
* @types/debug                                      
Unused devDependencies                              
* @types/jest                                                                                            
* @typescript-eslint/eslint-plugin                                                                       
* @typescript-eslint/parser                                                                              
* eslint-config-prettier                                                                                 
* eslint-plugin-import                              
* eslint-plugin-jest                                
* eslint-plugin-simple-import-sort                                                                                                                                                                                 
* eslint_d                                          
* ts-node                                           
psql: error: connection to server on socket "/run/postgresql/.s.PGSQL.5432" failed: No such file or directory
        Is the server running locally and accepting connections on that socket?
error Command failed with exit code 2.
```

However, `docker compose exec app yarn jest -i` passes, except for the test `providing just PGHOST and PGDATABASE is possible`, which fails on a clean checkout.

### CODE REVIEW

**Comment by octavifs** [comment]
+1 to the feature. We'd relly benefit at work from being able to specify a job_key on cron jobs. I would be happy to help, if possible.

File: __tests__/cron-timing.test.ts [hunk 1]
```diff
@@ -139,3 +139,43 @@ test("clock skew doesn't prevent task from being scheduled at the right time", (
...
+    await cronFinishedBackfilling;
+    await poolReady;
+
+    const cronScheduleComplete = eventMonitor.awaitNext("cron:scheduled");
+
+    const cronScheduleCalls = eventMonitor.count("cron:schedule");
+    // Allow the first copy of the job to get scheduled
+    await setTime(REFERENCE_TIMESTAMP + 4 * HOUR + 1 * SECOND); // 4:00:01am
+    expect(cronScheduleCalls.count).toEqual(1);
+    let jobs = await getJobs(pgPool);
```
Comment (benjie) [inline __tests__/cron-timing.test.ts:165]: Let's check the jobs match

```suggestion
    const jobs = await getJobs(pgPool);
```

File: __tests__/cron-timing.test.ts [hunk 2]
```diff
@@ -139,3 +139,43 @@ test("clock skew doesn't prevent task from being scheduled at the right time", (
...
+      expect.objectContaining({ task_identifier: "my_task", key: "foo" }),
+    ]);
+
+    // Allow the system to reschedule the job after seeing it wasn't picked up
+    await setTime(REFERENCE_TIMESTAMP + 8 * HOUR + 1 * SECOND); // 8:00:01am
+    expect(cronScheduleCalls.count).toEqual(2);
+    jobs = await getJobs(pgPool);
+    expect(jobs).toEqual([
+      expect.objectContaining({ task_identifier: "my_task", key: "foo" }),
+    ]);
```
Comment (benjie) [inline __tests__/cron-timing.test.ts:176]: Not sure if this is quite the right assertion, but something like this:

```suggestion
    const jobs2 = await getJobs(pgPool);
    expect(jobs2).toEqual([
      expect.objectContaining({ task_identifier: "my_task", key: "foo", id: jobs[0].id }),
    ]);
```

File: src/cronConstants.ts [hunk 3]
```diff
@@ -39,6 +39,10 @@ export const CRONTAB_OPTIONS_BACKFILL = /^((?:[0-9]+[smhdw])+)$/;
 export const CRONTAB_OPTIONS_MAX = /^([0-9]+)$/;
 /** Matches the queue=name option, capturing the queue name */
 export const CRONTAB_OPTIONS_QUEUE = /^([-a-zA-Z0-9_:]+)$/;
+/** Matches the jobKey=foo option, capturing the unique identifier for the job */
+export const CRONTAB_OPTIONS_JOB_KEY = /^(.*)$/;
+/** Matches the jobKeyMode=replace option, defining the replacement strategy this job */
+export const CRONTAB_OPTIONS_JOB_KEY_MODE = /^(replace|preserve_run_at)$/;
```
Comment (benjie) [inline src/cronConstants.ts:45]: Should we also allow `unsafe_dedupe`?

**Review by benjie — CHANGES_REQUESTED** [review]
This needs to be reworked on top of the latest Worker because I changed a number of internals, but I'm generally in favour of it - great work!

And sorry it's taken so long to get around to, I've been more than a little overwhelmed with work recently 😅

**Comment by you-fail-me** [comment]
Very much looking forward for this to get merged, would really help me a lot in prod

**Comment by a613** [comment]
This is a pretty critical omission as cronjobs are, by their fixed interval nature, prone to running into each other if the work takes too long. Can we possibly get some attention on this?

**Review by benjie — APPROVED** [review]
Thanks for the great work @spiffytech :raised_hands: 

I've merged the latest, applied my suggestions, added a couple additional assertions to the test, and tidied up parsing of the opts string to avoid being too noisy. This is good to go now!

**Review by benjie — APPROVED** [review]

**Pull request merged**